source/server/options_impl_platform_linux.cc

```diff
@@ -0,0 +1,49 @@
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <bitset>
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <thread>
+
+#include "server/options_impl_platform.h"
+
+namespace Envoy {
+
+uint32_t OptionsImplPlatform::getCpuCount() {
+  unsigned int hw_threads = std::thread::hardware_concurrency();
+  std::string path("/proc/" + std::to_string(getpid()) + "/status");
+  std::string key("Cpus_allowed:");
+  std::ifstream file(path);
+  if (!file) {
+    // Fall back to number of hardware threads.
+    return hw_threads;
+  }
+
+  std::string statusLine;
+  unsigned int threads = 0;
+
+  while (std::getline(file, statusLine)) {
+    if (statusLine.compare(0, key.size(), key, 0, key.size()) == 0) {
+      std::string value = statusLine.substr(key.size());
+      for (std::string::iterator iter = value.begin(); iter != value.end(); iter++) {
+        if (std::isxdigit(*iter)) {
+          char buf[2];
+          buf[0] = *iter;
+          buf[1] = '\0';
+          std::bitset<4> cpus(strtol(buf, NULL, 16));
+          threads += cpus.count();
+        }
+      }
+      break;
+    }
+  }
+  // Sanity check.
+  if (threads > 0 && threads <= hw_threads) {
+    return threads;
+  }
+  return hw_threads;
```

Comment:
This might mean we return 0, which may surprise an unsuspecting caller?

Reply:
You're right. I added a check to make sure `hw_threads >= 1`.